### ISSUE TITLE
Support for MariaDB 10.4

### ISSUE THREAD

**Issue by UnleavenedEagle** [virtualmin/virtualmin-gpl#95]
Virtualmin Version: 6.07
MariaDB version: 10.4.6

Error occurs when creating user from Virtualmin who has permission to a database, or after modifying access to database.

`Failed to save user : SQL insert into user (host, user, ssl_type, ssl_cipher, x509_issuer, x509_subject, plugin, authentication_string) values ('localhost', 'TestUser', '', '', '', '', 'mysql_native_password', '*0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZTEST') failed : The target table user of the INSERT is not insertable-into`

MariaDB has now a `CREATE USER` function.
https://mariadb.com/kb/en/library/create-user/

**Comment by jcameron** [comment]
This maybe caused by a mis-identification of the MySQL version.

What does the command `select version()` output when run in MySQL on your system?

**Comment by mzglinski** [comment]
Unfortunately, this is caused by this change: https://mariadb.com/kb/en/library/mysqluser-table/ 
(MariaDB starting with 10.4 section)

**Comment by UnleavenedEagle** [comment]
> What does the command `select version()` output when run in MySQL on your system?

```
+------------------------------------------+
| version()                                |
+------------------------------------------+
| 10.4.6-MariaDB-1:10.4.6+maria~bionic-log |
+------------------------------------------+
1 row in set (0.001 sec)
```

**Comment by jcameron** [comment]
Thanks - we'll work on fixing Virtualmin to work in this case.

**Comment by lofidev** [comment]
I'm in the same situation as @UnleavenedEagle, select version() output:
```
+----------------+
| VERSION()      |
+----------------+
| 10.4.6-MariaDB |
+----------------+
1 row in set (0.000 sec)
```

Rolling back Virtualmin as suggested in the [forums](https://www.virtualmin.com/node/64694) sadly did not fix the issue for me.

**Comment by James-Oakley** [comment]
See my issue report on the VM site at https://www.virtualmin.com/node/66717

Basically, there is code to identify when we're using MariaDB 10.2 or later, and to avoid using authentication_string in those cases. However, that switch case also requires the variable $plainpass to be non-empty, which is not always the case. So it therefore uses the old authentication_string method. In MariaDB 10.2 and 10.3, this still worked (deprecated) but support for it has now been dropped in 10.4 which is where the problems are coming from.

**Comment by jcameron** [comment]
For those running MariaDB 10.4 - which Linux distribution/version are you running, and is this a package supplied with the distro or did you install it separately?

**Comment by James-Oakley** [comment]
In my case, Centos 7. I didn't exactly install MariaDB separately (I still let the VirtualMin installer start work on a minimal system), but I forced it to use MariaDB 10.4 by adding their repository to /etc/yum.repos.d before running the VirtualMin installer.

**Comment by UnleavenedEagle** [comment]
I am using Ubuntu Bionic LTS, and I installed MariaDB by adding the repository from the [repository configuration tool](https://downloads.mariadb.org/mariadb/repositories/#mirror=ukfast) on MariaDB.org. So not from the default repositories, as that only included an older version of MariaDB (10.1 I think).

**Comment by lofidev** [comment]
Similar to @James-Oakley I'm using Centos 7.6.1810 and installed from: /etc/yum.repos.d/MariaDB10.repo

**Comment by rostovtsev** [comment]
Anyone seeing this issue, could you please apply the patch and see if it works for you?

Webmin needs to be restarted after applying this patch.

https://github.com/virtualmin/virtualmin-gpl/commit/b414d655de88675fa2cb715ec81875df2ffaa537

**Comment by mzglinski** [comment]
Assigning a user to database works with this patch (deleting users also works). Should this patch also fix modifying 'Remote Hosts' for all users within the virtual server?

**Comment by UnleavenedEagle** [comment]
I am sorry if I am being stupid here, but how do I apply the patch? I have ran a search for the file feature-mysql.pl on the machine Virtualmin is installed on, and it can't find such file. Perl isn't a compiled language, so the problem is not that this is only a source file, and I have to compile it, but I have a feeling that it will be something similar to that.

**Comment by UnleavenedEagle** [comment]
Sorry if I am just being a bit silly here, but I can't seem to install this patch. I can't seem to find the feature-mysql.pl file. I have two suspisions, one is that that's only a source file, but I highly doubt that as Perl is an interpreted language, although I have no experience with Perl; my second suspicion is that it's inside a virtual-server.wbm or virtualmin.wbm file which is really just a tar.gz file with a different extension, but I can't seem to be able to find that file. Might it be missing because I've installed Virtualmin via the .deb installer?

**Comment by mzglinski** [comment]
@UnleavenedEagle for me the file was located at `/usr/share/webmin/virtual-server/feature-mysql.pl`

@rostovtsev I think that remote hosts functionality is not working due to these lines: https://github.com/virtualmin/virtualmin-gpl/blob/b414d655de88675fa2cb715ec81875df2ffaa537/feature-mysql.pl#L2290

However, I'm unfamiliar with Perl and unfortunately cannot propose the fix.

**Comment by rostovtsev** [comment]
> I don't really know how to install the patch.

Replace you local file that is located under either _/usr/libexec/webmin/virtual-server_ or _/usr/share/webmin/virtual-server_ with [patched one](https://raw.githubusercontent.com/virtualmin/virtualmin-gpl/master/feature-mysql.pl) and restart Webmin by running in console:

```
/etc/webmin/restart
```

**Comment by jcameron** [comment]
So we've made some progress here in improving support for MariaDB 10.2, but 10.4 is still underway.

**Comment by UnleavenedEagle** [comment]
The patch works for me.

**Comment by jcameron** [comment]
Ok, the next release of Virtualmin and Webmin will support user management on all MariaDB versions.

**Comment by iliarostovtsev** [comment]
Complete support for MariaDB 10.4+ and MySQL 8+ has been implemented in Webmin 1.960.
